src/mainboard/starfive/visionfive1/bt0/src/main.rs

```diff
@@ -0,0 +1,249 @@
+#![feature(naked_functions, asm_sym, asm_const)]
+#![no_std]
+#![no_main]
+
+use core::{
+    arch::asm,
+    intrinsics::transmute,
+    panic::PanicInfo,
+    ptr::{read_volatile, write_volatile},
+    slice,
+};
+use embedded_hal::serial::nb::Write;
+use init::{clock_init, iopad_init, rstgen_init};
+use riscv::register::{marchid, mhartid, mimpid, mvendorid};
+
+mod dram;
+mod dram_lib;
+mod dram_phy;
+mod dram_pi;
+mod dram_pi_start;
+mod init;
+#[macro_use]
+mod log;
+
+// see SiFive VICU7 manual chapter 6 (p 31)
+const CLINT_BASE_ADDR: usize = 0x0200_0000;
+const CLINT_HART1_MSIP: usize = CLINT_BASE_ADDR + 0x0004;
+// see JH7100 datasheet
+const SRAM0_BASE: usize = 0x1800_0000;
+const SRAM0_SIZE: usize = 0x0002_0000;
+const DRAM_BASE: usize = 0x8000_0000;
+
+const MAIN_BLOB_BASE: usize = SRAM0_BASE + 30 * 1024;
+const MAIN_BLOB_SIZE: usize = 2 * 1024;
+const SRAM1_BASE: usize = 0x1808_0000;
+const SPI_FLASH_BASE: usize = 0x2000_0000;
+const DRAM_BLOB_BASE: usize = SPI_FLASH_BASE + 0x0001_0000;
+const PAYLOAD_BASE: usize = SPI_FLASH_BASE + 0x0004_0000;
+const DTB_ADDR: usize = DRAM_BASE + 0x0020_0000 + 0x0100_0000; // TODO
+const LOAD_ADDR: usize = DRAM_BASE;
+const LOAD_MAIN: bool = false;
+const DEBUG: bool = false;
+
+const STACK_SIZE: usize = 4 * 1024; // 4KiB
+
+const QSPI_CSR: usize = 0x1186_0000;
+const QSPI_READ_CMD: usize = QSPI_CSR + 0x0004;
+const SPI_FLASH_READ_CMD: u32 = 0x0003;
+
+pub type EntryPoint = unsafe extern "C" fn(r0: usize, dtb: usize);
+
+#[link_section = ".bss.uninit"]
+static mut BT0_STACK: [u8; STACK_SIZE] = [0; STACK_SIZE];
+
+/// Set up stack and jump to executable code.
+///
+/// # Safety
+///
+/// Naked function.
+#[naked]
+#[export_name = "start"]
+#[link_section = ".text.entry"]
+#[allow(named_asm_labels)]
+pub unsafe extern "C" fn start() -> ! {
+    asm!(
+        // Clear feature disable CSR
+        "csrwi  0x7c1, 0",
```

Comment:
This CSR does not exist on visionfive1, it's a T-Head dedicated one which can be found on D1's C906. I suspect we should remove this line of code to have visionfive1 bootloaded.

Reply:
Are you sure? It's SiFive's "feature disable CSR" as described in the U74 core manual and also in the vendor's code:
![Screenshot_20221222-120653_Document_Viewer](https://user-images.githubusercontent.com/4245199/209121515-80eb11b2-25fc-420e-a070-72c82ecb6a6b.png)


https://github.com/starfive-tech/JH7100_secondBoot/blob/e17302063c9a4b74475b18ff24dd149c27257354/boot/start.S#L32

Reply:
@orangecms: Oh sorry! I forgot about that. SiFive also have this type of registers, so we should keep this line of code